# Analysis of Auto Trade Classifieds in the platform of Ebay Kleinanzeigen (German Ebay portal for small classifieds)

In [1]:
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv', encoding ='Latin-1')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
info_autos = autos.info()
TOP_5_autos = autos.head(5)
LAST_5_autos = autos.tail(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

### The DS contains 20 cols, being most of them strings (15) and only 5 integers. And 50.000 rows!
### From a first glympse we can tell that there are labels that need to be renamed, e g, yearOfRegistration, offerType, nrOfPictures. They are mainly written using CamelCase, instead of Pythons preferred snake_case. Columns that need to be retyped, e g, price. And columns that need to be completed due to the lack of values, e g, notRepairedDamage, fuelType, model, gearbox. These cols present cases of null values, not much, but they do.
### We need to do some basic Data Cleaning in the DS

In [4]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [5]:
autos.rename(columns={'dateCrawled': 'date_crawled', 'name': 'name', 'seller':'seller', 'offerType':'offer_type', 'price': 'price', 'abtest': 'abtest',
       'vehicleType': 'vehicle_type', 'yearOfRegistration': 'registration_year', 'gearbox': 'gearbox', 'powerPS': 'power_ps', 'model': 'model',
       'odometer': 'odometer', 'monthOfRegistration': 'registration_month', 'fuelType': 'fuel_type', 'brand': 'brand',
       'notRepairedDamage': 'unrepaired_damage', 'dateCreated': 'ad_created', 'nrOfPictures': 'pics_number', 'postalCode': 'postal_code',
       'lastSeen': 'last_seen'}, inplace=True)

In [6]:
autos.head(5)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  pics_number  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

### We just renamed all the cols, and changed them to a more Pythonic readable way. Changing them from CamelCase to Snake_case.
### Doing so we are not only harmonizing our DS but also preventing us from making any typo related mistakes.

In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-04 16:40:33  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  pics_number   postal_code            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

In [8]:
autos['pics_number'].value_counts()

0    50000
Name: pics_number, dtype: int64

#### The column *pics_number* is populated only with the value 0. Therefore, there is no analytical value in having this column in our DataFrame (DF)

In [9]:
autos['seller'].value_counts()



privat        49999
gewerblich        1
Name: seller, dtype: int64

#### The column *seller* has also no relevance in our DF, since all values point into the direction of private sales, and only one datapoint appear to be a comercial add. We can drop this column. 

In [10]:
autos['offer_type'].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

#### The same applies to *offer_type* col

In [11]:
autos['price'].value_counts()

$0          1421
$500         781
$1,500       734
$2,500       643
$1,200       639
$1,000       639
$600         531
$3,500       498
$800         498
$2,000       460
$999         434
$750         433
$900         420
$650         419
$850         410
$700         395
$4,500       394
$300         384
$2,200       382
$950         379
$1,100       376
$1,300       371
$3,000       365
$550         356
$1,800       355
$5,500       340
$1,250       335
$350         335
$1,600       327
$1,999       322
            ... 
$11,360        1
$11,825        1
$36,800        1
$6,940         1
$7,985         1
$34,996        1
$35,990        1
$43,461        1
$6,895         1
$1,098         1
$15,749        1
$22,555        1
$4,335         1
$1,759         1
$2,895         1
$51,999        1
$960           1
$17,149        1
$13,387        1
$2,479         1
$6,295         1
$57,800        1
$3,485         1
$94,999        1
$265,000       1
$5,485         1
$46,200        1
$15,480       

#### The column *price* is a int64 type, but it has symbols in it and not only numbers. It needs to be cleaned!

In [12]:
autos['odometer'].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

#### Same thing applies to the *odometer* col. Although it is an integer type column we still have to deal with symbols in it (km). A cleansing need to be done!

In [13]:
print(autos['abtest'].head(10))

0    control
1    control
2       test
3    control
4       test
5       test
6       test
7    control
8       test
9    control
Name: abtest, dtype: object


In [14]:
autos['power_ps'].max()

17700

#### The *power_ps* col has for sure an error in its max data point. Its retreiving the value of 17700. No car has that many horse power! Needs to be looked into and solved!

In [15]:
autos['registration_month'].min()

0

In [16]:
autos['registration_month'].value_counts()

0     5075
3     5071
6     4368
5     4107
4     4102
7     3949
10    3651
12    3447
9     3389
11    3360
1     3282
8     3191
2     3008
Name: registration_month, dtype: int64

#### This column presents an abnormal number of 0's for a registration month. They should begin in 1 (first month of the year), and end in 12 (the last month). So more deepth into this trend should be taken care!

In [17]:
autos['ad_created'].head(10)

0    2016-03-26 00:00:00
1    2016-04-04 00:00:00
2    2016-03-26 00:00:00
3    2016-03-12 00:00:00
4    2016-04-01 00:00:00
5    2016-03-21 00:00:00
6    2016-03-20 00:00:00
7    2016-03-16 00:00:00
8    2016-03-22 00:00:00
9    2016-03-16 00:00:00
Name: ad_created, dtype: object

In [18]:
autos['postal_code'].head()

0    79588
1    71034
2    35394
3    33729
4    39218
Name: postal_code, dtype: int64

In [19]:
autos['registration_year'].min()

1000

In [20]:
autos['registration_year'].max()

9999

In [21]:
autos['registration_year'].value_counts()

2000    3354
2005    3015
1999    3000
2004    2737
2003    2727
2006    2708
2001    2703
2002    2533
1998    2453
2007    2304
2008    2231
2009    2098
1997    2028
2011    1634
2010    1597
2017    1453
1996    1444
2012    1323
2016    1316
1995    1313
2013     806
2014     666
1994     660
2018     492
1993     445
2015     399
1990     395
1992     391
1991     356
1989     181
        ... 
1950       3
1955       2
9000       2
1954       2
1800       2
1957       2
1941       2
1951       2
1934       2
4100       1
4800       1
1953       1
1111       1
1927       1
6200       1
4500       1
1943       1
5911       1
1939       1
1938       1
2800       1
8888       1
1000       1
1500       1
1948       1
1931       1
1929       1
1001       1
9996       1
1952       1
Name: registration_year, Length: 97, dtype: int64

#### In the col *registration_year* we have a similar problem as in the col *registration_month*. But here both on the max and min values (9999 and 1000). This data for sure is wrong. It can't be possible for a car to be registered on the year 1000 as in the year 9999!

### Both *price* and *odometer* are numeric cols saved as text in our DF, and with symbols in them. What we are goint to do next is clean this cols, convert them to a numeric type, and rename them

In [22]:
# cleaning and converting the price col:
autos['price'] = autos['price'].str.replace('$', '').str.replace(',','').astype('int64')

In [23]:
autos['price'].head(3)

0    5000
1    8500
2    8990
Name: price, dtype: int64

In [24]:
# cleaning and converting the odometer col:
autos['odometer'] = autos['odometer'].str.replace('km', '').str.replace(',','').astype('int64')

In [25]:
autos.rename(columns={'odometer':'odometer_km'}, inplace=True)

In [26]:
autos['odometer_km'].head(3)

0    150000
1    150000
2     70000
Name: odometer_km, dtype: int64

#### Investigating deeper these 2 cols (*price* and *odometer_km*) and looking for outliers in both of them

In [27]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [28]:
#autos['price'].unique().shape
#autos['price'].describe()
#autos['price'].max()
#autos['price'].min()
autos['price'].value_counts().sort_index(ascending=False).head(20)
#autos['odometer_km'].describe()
#autos['odometer_km'].value_counts().sort_index(ascending=True).tail(20)


99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [29]:
autos['price'].value_counts().sort_index(ascending=True).head(50)

0      1421
1       156
2         3
3         1
5         2
8         1
9         1
10        7
11        2
12        3
13        2
14        1
15        2
17        3
18        1
20        4
25        5
29        1
30        7
35        1
40        6
45        4
47        1
49        4
50       49
55        2
59        1
60        9
65        5
66        1
70       10
75        5
79        1
80       15
89        1
90        5
99       19
100     134
110       3
111       2
115       2
117       1
120      39
122       1
125       8
129       1
130      15
135       1
139       1
140       9
Name: price, dtype: int64

In [30]:
autos[autos['price'] >= 999990].head(9)

date_crawled                                               name  \
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
37585  2016-03-29 11:38:54                                Volkswagen_Jetta_GT   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   

       seller offer_type     price   abtest vehicle_type  registration_year  \
514    privat    Angebot    999999     test        kombi               2009   
2897   privat    Angebot  11111111     test    limousine               1973   
7814   privat    Angebot   1300000  control        coupe               1992   
11137  privat    Angebot  10000000  control        coupe               1960   
22947  privat    Angebot   1234566  control        kombi               1999   
24384  privat    Angebot  11111111     test          NaN               1995   
27371  privat    Angebot  12345678  control          NaN               2017   
37585  privat    Angebot    999990     test    limousine               1985   
39377  privat    Angebot  12345678  control          NaN               2018   

         gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
514      manuell       101   focus       125000                   4    benzin   
2897     manuell        48  escort        50000                   3    benzin   
7814         NaN         0     NaN        50000                  12       NaN   
11137    manuell       368     NaN       100000                   1    benzin   
22947  automatik       190     NaN       150000                   2    diesel   
24384        NaN         0     NaN       150000                   0       NaN   
27371        NaN        95   punto       150000                   0       NaN   
37585    manuell       111   jetta       150000                  12    benzin   
39377    manuell        95     v40       150000                   6       NaN   

                brand unrepaired_damage           ad_created  pics_number  \
514              ford              nein  2016-03-17 00:00:00            0   
2897             ford              nein  2016-03-12 00:00:00            0   
7814   sonstige_autos              nein  2016-04-04 00:00:00            0   
11137  sonstige_autos              nein  2016-03-29 00:00:00            0   
22947             bmw               NaN  2016-03-22 00:00:00            0   
24384      volkswagen               NaN  2016-03-21 00:00:00            0   
27371            fiat               NaN  2016-03-09 00:00:00            0   
37585      volkswagen                ja  2016-03-29 00:00:00            0   
39377           volvo              nein  2016-03-08 00:00:00            0   

       postal_code            last_seen  
514          12205  2016-04-06 07:17:35  
2897         94469  2016-03-12 22:45:27  
7814         60598  2016-04-05 11:34:11  
11137        73033  2016-04-06 21:18:11  
22947        17454  2016-04-02 03:17:32  
24384        18519  2016-03-21 14:40:18  
27371        96110  2016-03-09 15:45:47  
37585        50997  2016-03-29 11:38:54  
39377        14542  2016-04-06 23:17:31

#### In the cols *odometer_km* we could not detect any incongruences/outliers in the data points. 
#### The same thing cant't be said about the col *price*. Here we could find several inconsistencies. Plenty of data points with the value 0 (#1421), as also some data points with unrealistic high price values for used cars (>999990). E.g., Fiat_Punto for 12345678€; Volkswagen_Jetta_GT for 999990€; Ford_Focus_Turnier_1.6_16V_Style for 999999€! 
#### What we decided to do was to define a cut-off for the col *price* and remove these outliers. Applying a filter (boolean indexing) to remove all the prices that stood between 99€ and no more than 999990€:

In [31]:
autos = autos[(autos['price'] < 999990) & (autos['price'] >= 99)]
#autos['price'] = autos[autos['price'].between(99,999990)]

In [32]:
autos['price'].describe()

count     48243.000000
mean       5928.074809
std        9077.322281
min          99.000000
25%        1250.000000
50%        3000.000000
75%        7499.000000
max      350000.000000
Name: price, dtype: float64

In [33]:
autos.shape

(48243, 20)

In [34]:
autos_date_crawled_dist = autos['date_crawled'].value_counts()

In [35]:
date_crawled = autos['date_crawled'].str[:10]

In [36]:
date_crawled.head()

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: date_crawled, dtype: object

In [37]:
date_crawled.value_counts(dropna=False, normalize=True).sort_index(ascending=True)

2016-03-05    0.025372
2016-03-06    0.014033
2016-03-07    0.036047
2016-03-08    0.033207
2016-03-09    0.033000
2016-03-10    0.032274
2016-03-11    0.032606
2016-03-12    0.036897
2016-03-13    0.015671
2016-03-14    0.036648
2016-03-15    0.034305
2016-03-16    0.029455
2016-03-17    0.031528
2016-03-18    0.012893
2016-03-19    0.034762
2016-03-20    0.037788
2016-03-21    0.037228
2016-03-22    0.032917
2016-03-23    0.032295
2016-03-24    0.029455
2016-03-25    0.031507
2016-03-26    0.032316
2016-03-27    0.031113
2016-03-28    0.034948
2016-03-29    0.034098
2016-03-30    0.033725
2016-03-31    0.031860
2016-04-01    0.033704
2016-04-02    0.035591
2016-04-03    0.038596
2016-04-04    0.036544
2016-04-05    0.013059
2016-04-06    0.003171
2016-04-07    0.001389
Name: date_crawled, dtype: float64

#### All Data was crawled between the beginning of month 03 (March) and month 04 (April) of the year 2016. 

In [38]:
ad_created = autos['ad_created'].str[:10]

In [39]:
ad_created.value_counts(dropna=False, normalize=True).sort_index(ascending=True)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033083
2016-03-10    0.031984
2016-03-11    0.032917
2016-03-12    0.036731
2016-03-13    0.017039
2016-03-14    0.035280
2016-03-15    0.034036
2016-03-16    0.029953
2016-03-17    0.031196
2016-03-18    0.013577
2016-03-19    0.033642
2016-03-20    0.037850
2016-03-21 

#### While all the ads crawled were created in the website between month 05 (May) of 2015 and month 04 (April) of 2016. So we have approximately 1 year of Data in our DataSet.
#### The day that collected more ads being 03-04-2016 with approximately 4% of all the ads crawled. 

In [40]:
ad_created_month = autos['ad_created'].str[:7]

In [41]:
ad_created_month.value_counts(dropna=False, normalize=True)

2016-03    0.837365
2016-04    0.160997
2016-02    0.001264
2016-01    0.000249
2015-12    0.000041
2015-06    0.000021
2015-09    0.000021
2015-11    0.000021
2015-08    0.000021
Name: ad_created, dtype: float64

#### The month with more data points in our DS is March 2016 (83,7%), followed by April (16,1%). 
#### This is related to the time when the Data was crawled, months 03 beginning of 04, and the rotation of the ads in the website - high!

In [42]:
last_seen = autos['last_seen'].str[:10]

In [43]:
last_seen.value_counts(dropna=False, normalize=True).sort_index(ascending=True)

2016-03-05    0.001078
2016-03-06    0.004312
2016-03-07    0.005431
2016-03-08    0.007317
2016-03-09    0.009597
2016-03-10    0.010634
2016-03-11    0.012396
2016-03-12    0.023775
2016-03-13    0.008872
2016-03-14    0.012624
2016-03-15    0.015878
2016-03-16    0.016438
2016-03-17    0.028087
2016-03-18    0.007317
2016-03-19    0.015795
2016-03-20    0.020666
2016-03-21    0.020563
2016-03-22    0.021350
2016-03-23    0.018593
2016-03-24    0.019775
2016-03-25    0.019132
2016-03-26    0.016666
2016-03-27    0.015546
2016-03-28    0.020853
2016-03-29    0.022283
2016-03-30    0.024708
2016-03-31    0.023817
2016-04-01    0.022843
2016-04-02    0.024895
2016-04-03    0.025123
2016-04-04    0.024542
2016-04-05    0.125054
2016-04-06    0.221939
2016-04-07    0.132102
Name: last_seen, dtype: float64

#### This col is ranged as the date_crawled col!

In [44]:
autos['registration_year'].describe()

count    48243.000000
mean      2004.729702
std         87.880224
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

#### From the descriptive analysis of this col we can observe that the average year of registration for the vehicles offered is 2004. But its also visible that we have multiple data points that are probably not correct. We have a min value of 1000 and a max value of 9999. This registration dates can't be right!

In [45]:
autos['registration_year'].value_counts(normalize=True).sort_index()

1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000041
1910    0.000041
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000041
1937    0.000083
1938    0.000021
1939    0.000021
1941    0.000041
1943    0.000021
1948    0.000021
1950    0.000021
1951    0.000041
1952    0.000021
1953    0.000021
1954    0.000041
1955    0.000041
1956    0.000083
1957    0.000041
1958    0.000083
1959    0.000124
1960    0.000456
1961    0.000124
1962    0.000083
1963    0.000166
1964    0.000249
          ...   
2000    0.064362
2001    0.054495
2002    0.051344
2003    0.055842
2004    0.055946
2005    0.060361
2006    0.055303
2007    0.047116
2008    0.045810
2009    0.043115
2010    0.032896
2011    0.033539
2012    0.027133
2013    0.016603
2014    0.013722
2015    0.007898
2016    0.024936
2017    0.028688
2018    0.009701
2019    0.000041
2800    0.000021
4100    0.000021
4500    0.000021
4800    0.000021
5000    0.000062
5911    0.000021
6200    0.000021
8888    0.0000

In [46]:
autos[autos['registration_year'] == 2017].head(10)

date_crawled                                               name  \
10   2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
65   2016-04-04 19:30:39                      Ford_Fiesta_zum_ausschlachten   
68   2016-04-03 17:36:59  Mini_cooper_s_clubman_/vollausstattung_/_Navi/...   
113  2016-04-03 14:58:29                              Golf_4_Anfaenger_auto   
197  2016-04-05 10:36:24                              VW_Polo_9N_an_Bastler   
253  2016-03-27 13:25:18             Ford_mondeo_Gas_anlage_mit_TÜV_04.2017   
348  2016-03-17 20:58:24  VW_Beetle_1.8Turbo_mit_Vollausstattung_und_seh...   
438  2016-03-10 20:36:25                VW_Golf_5_V_1.6_2004_Klima_Tempomat   
454  2016-03-25 09:57:03    Ford_Fiesta_Servo_D3_Kat_86€_Steuern_65.000_tkm   
457  2016-03-15 17:48:39  Mercedes_Benz_W203_C200_Kombi_Kompressor_c_kla...   

     seller offer_type  price   abtest vehicle_type  registration_year  \
10   privat    Angebot    999     test          NaN               2017   
65   privat    Angebot    250  control          NaN               2017   
68   privat    Angebot  10990     test          NaN               2017   
113  privat    Angebot   1200     test          NaN               2017   
197  privat    Angebot    888  control          NaN               2017   
253  privat    Angebot   2250     test          NaN               2017   
348  privat    Angebot   3750  control          NaN               2017   
438  privat    Angebot   4150     test          NaN               2017   
454  privat    Angebot    750     test          NaN               2017   
457  privat    Angebot   3000  control          NaN               2017   

     gearbox  power_ps     model  odometer_km  registration_month fuel_type  \
10   manuell        90       NaN       150000                   4    benzin   
65   manuell        65    fiesta       125000                   9    benzin   
68   manuell       174   clubman       100000                   0       NaN   
113  manuell        75      golf       150000                   7       NaN   
197  manuell        64      polo        20000                   7       NaN   
253  manuell         0    mondeo       150000                   8    benzin   
348  manuell       150    beetle       150000                   7       NaN   
438  manuell       102      golf       150000                   0    benzin   
454  manuell        50    fiesta        70000                   0    benzin   
457  manuell       163  c_klasse       150000                   8       NaN   

             brand unrepaired_damage           ad_created  pics_number  \
10      volkswagen              nein  2016-03-14 00:00:00            0   
65            ford               NaN  2016-04-04 00:00:00            0   
68            mini              nein  2016-04-03 00:00:00            0   
113     volkswagen               NaN  2016-04-03 00:00:00            0   
197     volkswagen                ja  2016-04-05 00:00:00            0   
253           ford              nein  2016-03-27 00:00:00            0   
348     volkswagen              nein  2016-03-17 00:00:00            0   
438     volkswagen               NaN  2016-03-10 00:00:00            0   
454           ford               NaN  2016-03-25 00:00:00            0   
457  mercedes_benz               NaN  2016-03-15 00:00:00            0   

     postal_code            last_seen  
10         86157  2016-04-07 03:16:21  
65         65606  2016-04-05 12:22:12  
68         83135  2016-04-05 17:26:26  
113        97656  2016-04-05 14:15:48  
197        58566  2016-04-07 13:16:13  
253        56575  2016-04-05 15:18:34  
348        45896  2016-03-24 17:17:50  
438        49377  2016-03-13 03:16:51  
454        12207  2016-03-26 19:47:15  
457        40627  2016-03-21 16:49:25

#### Because our data was crawled during the years of 2015-2016 we cannot have registration years in our ads that goes beyond the year of 2016. Nor registrations prior to 1900 (the first car was invented around 1885/1886)
#### Although the amount of data representing the *registration_year* of 2017 is still meaningful, accruing to approx 3%, we still have to remove this data points because they mislead our analysis with false conclusions!
#### Data prior to the *registration_year* of 1900 is not so significant!
#### We now are going to assume a cut-off for our *registration_year* col between 1900-2016 and filter our data:

In [47]:
autos = autos[(autos['registration_year'] >= 1900) & (autos['registration_year'] <= 2016)]

In [48]:
autos['registration_year'].value_counts(normalize=True).sort_index(ascending=True)

1910    0.000043
1927    0.000022
1929    0.000022
1931    0.000022
1934    0.000043
1937    0.000086
1938    0.000022
1939    0.000022
1941    0.000043
1943    0.000022
1948    0.000022
1950    0.000022
1951    0.000043
1952    0.000022
1953    0.000022
1954    0.000043
1955    0.000043
1956    0.000086
1957    0.000043
1958    0.000086
1959    0.000129
1960    0.000474
1961    0.000129
1962    0.000086
1963    0.000173
1964    0.000259
1965    0.000367
1966    0.000474
1967    0.000561
1968    0.000561
          ...   
1987    0.001553
1988    0.002868
1989    0.003688
1990    0.007160
1991    0.007289
1992    0.007915
1993    0.009058
1994    0.013522
1995    0.025749
1996    0.029265
1997    0.041600
1998    0.050550
1999    0.062088
2000    0.066961
2001    0.056696
2002    0.053418
2003    0.058098
2004    0.058206
2005    0.062799
2006    0.057537
2007    0.049019
2008    0.047660
2009    0.044857
2010    0.034225
2011    0.034893
2012    0.028229
2013    0.017274
2014    0.0142

#### Checking the unique values in the *brand* col:

In [49]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [50]:
autos['brand'].value_counts().head(20)

volkswagen        9803
bmw               5109
opel              4975
mercedes_benz     4480
audi              4022
ford              3237
renault           2184
peugeot           1384
fiat              1191
seat               846
skoda              761
nissan             711
mazda              706
smart              658
citroen            652
toyota             593
hyundai            465
sonstige_autos     442
volvo              423
mini               408
Name: brand, dtype: int64

### VW comes highlighted at number 1 being by far the brand with more cars advertised in Ebay Kleinenanzeigen, with 9803 ads. Followed by BMW with 5109 ads.
### Closing top 3 with most ads comes the brand Opel, with 4975 offers. Closely followed by Mercedes with 4480.
### The last manufacturer from the top 20 brands in our DataSet is Mini. Offering a scarce 408 cars for sale!

In [51]:
top_20_brands_mean_price = {}
top_20 = autos['brand'].value_counts().head(20)

for b in top_20.index:
    autos_top_20 = autos[autos['brand'] == b]
    avg_price_top_20 = round(autos_top_20['price'].mean(),2)
    top_20_brands_mean_price[b] = avg_price_top_20

print(sorted(top_20_brands_mean_price.items(), key=lambda kv: kv[1]))

[('renault', 2493.88), ('fiat', 2827.68), ('opel', 3003.16), ('peugeot', 3113.86), ('smart', 3596.4), ('ford', 3779.27), ('citroen', 3790.59), ('mazda', 4129.77), ('seat', 4433.42), ('nissan', 4756.66), ('volvo', 4993.21), ('toyota', 5167.09), ('hyundai', 5399.65), ('volkswagen', 5434.77), ('skoda', 6409.61), ('bmw', 8378.43), ('mercedes_benz', 8672.65), ('audi', 9380.72), ('mini', 10639.45), ('sonstige_autos', 12784.56)]


#### Our top 5 pricey from the 20 most common brands in our ads are: 
    Mini - 10639.45€; 
    Audi - 9380.72€; 
    Mercedes Benz - 8672.65€; 
    BMW - 8378.43€; 
    Skoda - 6409.61€
#### The bottom 5 from the 20 most common brands in our ads are are:
    Smart - 3596.4€;
    Peugeot - 3113.86€;
    Opel - 3003.16€;
    Fiat - 2827.68€;
    Renault - 2493.88€

#### Now we are going to aggregate our most common 20 brands in our ads by their average (avg) mileage in order to check if there is any relation between avg mileage and avg price:

In [52]:
top_20_brands_mean_mileage = {}

for b in top_20.index:
    autos_top_20 = autos[autos['brand'] == b]
    avg_mileage_top_20 = round(autos_top_20['odometer_km'].mean(),2)
    top_20_brands_mean_mileage[b] = avg_mileage_top_20

print(top_20_brands_mean_mileage)

{'seat': 121536.64, 'skoda': 110906.7, 'citroen': 119601.23, 'toyota': 115944.35, 'peugeot': 127127.89, 'hyundai': 106978.49, 'bmw': 132673.71, 'opel': 129395.98, 'mercedes_benz': 131025.67, 'mini': 88308.82, 'fiat': 117019.31, 'audi': 129245.4, 'mazda': 124553.82, 'volvo': 138581.56, 'ford': 124277.11, 'nissan': 118326.3, 'volkswagen': 128803.43, 'smart': 99734.04, 'renault': 128301.28, 'sonstige_autos': 90395.93}


#### Combining both dictionaries, avg_price and avg_mileage, into 1 DataFrame in order to facilitate their comparison and any links between both of them:

In [53]:
mileage_top_20_series = pd.Series(top_20_brands_mean_mileage)

In [54]:
price_top_20_series = pd.Series(top_20_brands_mean_price)

In [55]:
mileage_price_top_20_df = pd.DataFrame(mileage_top_20_series, columns=['top_20_brands_mean_mileage'])
mileage_price_top_20_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 20 entries, audi to volvo
Data columns (total 1 columns):
top_20_brands_mean_mileage    20 non-null float64
dtypes: float64(1)
memory usage: 320.0+ bytes


In [56]:
mileage_price_top_20_df.head()

top_20_brands_mean_mileage
audi                      129245.40
bmw                       132673.71
citroen                   119601.23
fiat                      117019.31
ford                      124277.11

In [57]:
mileage_price_top_20_df['top_20_brands_mean_price'] = price_top_20_series

In [58]:
mileage_price_top_20_df.head(20)

top_20_brands_mean_mileage  top_20_brands_mean_price
audi                             129245.40                   9380.72
bmw                              132673.71                   8378.43
citroen                          119601.23                   3790.59
fiat                             117019.31                   2827.68
ford                             124277.11                   3779.27
hyundai                          106978.49                   5399.65
mazda                            124553.82                   4129.77
mercedes_benz                    131025.67                   8672.65
mini                              88308.82                  10639.45
nissan                           118326.30                   4756.66
opel                             129395.98                   3003.16
peugeot                          127127.89                   3113.86
renault                          128301.28                   2493.88
seat                             121536.64                   4433.42
skoda                            110906.70                   6409.61
smart                             99734.04                   3596.40
sonstige_autos                    90395.93                  12784.56
toyota                           115944.35                   5167.09
volkswagen                       128803.43                   5434.77
volvo                            138581.56                   4993.21

### The relation between avg_mileage and avg_price is obvious, and we can explain why mini used cars are the most expensive ones although being a middle price range brand!
### Mini is the brand with less mileage on its used cars, an avg of 88308.82 km's. While top brands like BMW, Mercedes Benz, and Audi are among the ones with more mileage on an avg. Scoring 132673.71; 131025.67; 129245.40 km's, respectively!
### This might explain in part why these 3 brands, although top ones, all come after mini in their avg_price, due to their high mileage numbers, and why Skoda, although an entry/middle price range brand comes in 5th with an avg_price of 6409.61€, perhaps sustained by their short avg_mileage, compared to our top 3 brands, with 110906.70 km's

### Now, while still working with our top 20 brands, let's try to investigate if there is any relation between the age of the cars, average price and  their average mileage!
#### First we will need to creat a column only for the number of years of our cars. All we have is their registration year, and the last time the ad has been seen. With this info we can obtain their number of years.

In [59]:
autos['last_seen'].describe()

count                   46370
unique                  36924
top       2016-04-07 06:17:27
freq                        8
Name: last_seen, dtype: object

In [60]:
autos['last_seen_year'] = autos['last_seen'].str.split('-').str[0].astype(int)

In [61]:
autos['last_seen_year'].head()

0    2016
1    2016
2    2016
3    2016
4    2016
Name: last_seen_year, dtype: int64

In [62]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
31     2016-03-14 16:53:09                          Noch_gut_erhaltenen_C_320   
32     2016-03-20 05:03:03                               Corsa_mit_TÜV_5.2016   
33     2016-03-15 12:25:26  Audi_A3_2.0_TDI_Sportback_DPF_Kupplung_defekt_...   
...                    ...                                                ...   
49968  2016-04-01 17:49:15                  Mercedes_Benz_190_D_2.5_Automatik   
49969  2016-03-17 18:49:02             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

In [63]:
autos['n_years'] = autos['last_seen_year'] - autos['registration_year']

In [64]:
autos['n_years'].describe()

count    46370.000000
mean        13.061096
std          7.127107
min          0.000000
25%          8.000000
50%         13.000000
75%         17.000000
max        106.000000
Name: n_years, dtype: float64

#### With our new column (*n_years*) created and added to our DataFrame (DF), we can now calculate the avg number of years for our top 20 brands, and assess the relationship between avg_years, avg_price and avg_mileage for our top 20 brands!
#### We can also check that the avg number of years for all the ads from our DF is around 13 years!

In [65]:
top_20_brands_mean_n_years = {}

for b in top_20.index:
    autos_top_20 = autos[autos['brand'] == b]
    avg_n_years_top_20 = round(autos_top_20['n_years'].mean(),2)
    top_20_brands_mean_n_years[b] = avg_n_years_top_20
    
print(top_20_brands_mean_n_years)

{'seat': 11.44, 'skoda': 9.45, 'citroen': 11.51, 'toyota': 11.66, 'peugeot': 12.22, 'hyundai': 8.8, 'bmw': 12.95, 'opel': 13.74, 'mercedes_benz': 13.95, 'mini': 8.06, 'fiat': 12.75, 'audi': 11.83, 'mazda': 13.04, 'volvo': 15.4, 'ford': 13.21, 'nissan': 12.54, 'volkswagen': 13.52, 'smart': 10.55, 'renault': 13.09, 'sonstige_autos': 22.0}


#### It's immediatly visible what's our number 1 brand in terms of aging: Volvo, with an average of 15 years (15.4).
#### At number 2 with and avg of almost 14 years we have Mercedes Benz (13.95) 
#### And closing our top 3 for the average number of years for our top 20 brands in Ebay Kleinenanzeige, we have Opel, with and avg of 13.74 years!
#### Is there any relationship between age and price, or between age, price, and mileage?? Let's find out!!

In [66]:
n_years_top_20_series = pd.Series(top_20_brands_mean_n_years)

In [67]:
mileage_price_top_20_df['top_20_brands_mean_n_years'] = n_years_top_20_series

In [68]:
mileage_price_top_20_df.head(20)

top_20_brands_mean_mileage  top_20_brands_mean_price  \
audi                             129245.40                   9380.72   
bmw                              132673.71                   8378.43   
citroen                          119601.23                   3790.59   
fiat                             117019.31                   2827.68   
ford                             124277.11                   3779.27   
hyundai                          106978.49                   5399.65   
mazda                            124553.82                   4129.77   
mercedes_benz                    131025.67                   8672.65   
mini                              88308.82                  10639.45   
nissan                           118326.30                   4756.66   
opel                             129395.98                   3003.16   
peugeot                          127127.89                   3113.86   
renault                          128301.28                   2493.88   
seat                             121536.64                   4433.42   
skoda                            110906.70                   6409.61   
smart                             99734.04                   3596.40   
sonstige_autos                    90395.93                  12784.56   
toyota                           115944.35                   5167.09   
volkswagen                       128803.43                   5434.77   
volvo                            138581.56                   4993.21   

                top_20_brands_mean_n_years  
audi                                 11.83  
bmw                                  12.95  
citroen                              11.51  
fiat                                 12.75  
ford                                 13.21  
hyundai                               8.80  
mazda                                13.04  
mercedes_benz                        13.95  
mini                                  8.06  
nissan                               12.54  
opel                                 13.74  
peugeot                              12.22  
renault                              13.09  
seat                                 11.44  
skoda                                 9.45  
smart                                10.55  
sonstige_autos                       22.00  
toyota                               11.66  
volkswagen                           13.52  
volvo                                15.40

### The relationship between these 3 inputs is pretty obvious!
### Now we have a better explanation why Mini comes taged with the highest avg price, less km's and the less aged of our top 20 brands with an average of 8 years!
### Skoda enters our top 5 pricey cars probably due to their mix of low age (9.45 years) and not so high mileage
### Volvo, despite its mix between high avg mileage and high avg years, doesn't come with the lowest price tag. This is probably better explained due to the fact that their car models aim a more medium/high segment, thus with higher price tags than Renault or Opel, for example!
### Regarding our top 3 brands, and the more comparable ones due to their nature in terms of segment, Mercedes Benz, BMW, and Audi, it's clear why Audi comes at second in terms of avg price. It is the brand both with less avg mileage and avg number of years, 129245.40 km's and 11.83 years. And although these 3 brands, from all our top 20 brands, are probably tagged with the higher prices when they are new, none of the 3 come at number 1 in the avg price for used cars! the reason being almost certainly the mix between high number of years and high mileage! 
### It is importante to point out that a more precise analysis would to be made if we had an average price of the cars as new, and compare not the most popular ones in our DF, but the ones from the same segment, with similar sell as new price tags!

### Next let's investigate if there is a difference in price between cars that have unrepaired damage and cars that don't have unrepaired damage not only for our top 20 brands, but for all the ads in our DataFrame. And if there is a difference, let's try to calculate it's magnitude!

##### But first let's replace any german words we might have in our col *unrepaired_damage* for english ones:

In [69]:
autos['unrepaired_damage'].value_counts(dropna=False)

nein    33776
NaN      8108
ja       4486
Name: unrepaired_damage, dtype: int64

In [70]:
autos['unrepaired_damage'] = autos['unrepaired_damage'].str.replace('nein', 'no')

In [71]:
autos['unrepaired_damage'] = autos['unrepaired_damage'].str.replace('ja', 'yes')

In [72]:
autos['unrepaired_damage'].value_counts(dropna=False)

no     33776
NaN     8108
yes     4486
Name: unrepaired_damage, dtype: int64

#### Calculating the avg price for both damaged and undamaged cars:

In [73]:
mean_damaged = round(autos[autos['unrepaired_damage'] == 'yes'].loc[:,'price'].mean(),2)

In [74]:
mean_undamaged = round(autos[autos['unrepaired_damage'] == 'no'].loc[:,'price'].mean(),2)

In [75]:
print(abs(mean_damaged - mean_undamaged))

4908.66


In [76]:
print(abs(mean_damaged - mean_undamaged)/(mean_undamaged)*100)

68.40098658082856


### We can see a huge price gap between damaged and undamaged cars, ca. 4909€, or 68.4% more on the price for the undamaged ones!

### Let's try now to compute the avg price of all our ads taking into consideration their age!

#### First we need to define our intervals for the number of years that the cars have. And then calculate the average for each interval.

#### Defining our intervals for the number of years of our advertised cars and the average price for each interval:

In [77]:
# avg_price_0-5:
round(autos[(autos['n_years'] >= 0) & 
                      (autos['n_years'] < 5)].loc[:,'price'].mean(),2)

15438.46

In [78]:
# avg_price_5-10:
round(autos[(autos['n_years'] >= 5) & 
                       (autos['n_years'] < 10)].loc[:,'price'].mean(),2)

10668.27

In [79]:
# avg_price_10-15:
round(autos[(autos['n_years'] >= 10) & 
                       (autos['n_years'] < 15)].loc[:,'price'].mean(),2)

4620.85

In [80]:
# avg_price_15-20: 
round(autos[(autos['n_years'] >= 15) & 
                       (autos['n_years'] < 20)].loc[:,'price'].mean(),2)

1961.17

In [81]:
# avg_price_+20: 
round(autos[autos['n_years'] >= 20].loc[:,'price'].mean(),2)

3410.89

### An interesting fact is that cars start to appreciate from age 20 onwards.
### The avg price for cars between age 0 and 5 is 15438.46€. That price tag drops approx 5000€ when our age range increases 5 years (5-10).
### The avg price for the cars aged between 10-15 is moreless 30% of the avg seen in the first range (0-5).
### And we can afirme that cars hit their residual avg value when they reach their interval age of 15-20.